### ISSUE TITLE
Sleep crashes inside OS threads.

### ISSUE THREAD

**Issue by Cazra** [racket/racket#4670]
**What version of Racket are you using?**
8.6 [cs]

**What program did you run?**

```
#lang racket/base

(require ffi/unsafe/os-thread)

(define sema (make-os-semaphore))
(call-in-os-thread (λ ()
  (sleep 1)
  (display "OK\n")
  (os-semaphore-post sema)))
(os-semaphore-wait sema)
```

**What should have happened?**
The OS thread should sleep for 1 second and display "OK".

**If you got an error message, please include it here.**
`invalid memory reference.  Some debugging context lost`

**Please include any other relevant details**
Operating system: Ubuntu 20.04.5 LTS

**Comment by mflatt** [comment]
Thanks for the report! But the bug here is an imprecision in the documentation of `call-in-os-thread`: `sleep` is in the same category as Racket thread and synchronization functions, and it won't work in the `call-in-os-thread` callback.

**Comment by Cazra** [comment]
It also looks like channel operations don't play well with os-threads either.

The following code:
```
(define sema (make-os-semaphore))
(define ch (make-channel))
(call-in-os-thread (λ ()
  (channel-put ch "OK")))
(os-semaphore-wait sema)

(define result (sync ch))
(display (format "Channel test result: ~a\n" result))
```

crashes with the error:
```
waiter-accessor: contract violation
  expected: waiter?
  given: #f
```

**Comment by mflatt** [comment]
Yes, that's right. Channels are about thread synchronization, so they're out. The `os-asynch-channel` API can help there, though, although as an asynchronous variant.

**Comment by Cazra** [comment]
Thanks! I'll try that out!

**Comment by Cazra** [comment]
Is there any way to kill an OS thread, like, an equivalent to `kill-thread` but for OS threads?

**Comment by LiberalArtist** [comment]
> Thanks for the report! But the bug here is an imprecision in the documentation of `call-in-os-thread`: `sleep` is in the same category as Racket thread and synchronization functions, and it won't work in the `call-in-os-thread` callback.

Maybe the confusion here is that the Racket function `sleep` sleeps in terms of Racket's green threads, so it's different from the C function named `sleep`, which sleeps in terms of OS threads. You can access C's `sleep` easily through the FFI, though, and, if that kind of `sleep`ing is what you want, it's probably much better than busy sleeping. Here's an example:

```racket
#lang racket
(require ffi/unsafe
         ffi/unsafe/define
         ffi/unsafe/os-thread
         ffi/unsafe/os-async-channel)
(define-ffi-definer define-libc #f)
(define-libc libc:sleep
  (_fun _uint -> _uint)
  #:c-id sleep)
(define (demo seconds)
  (define os-ach (make-os-async-channel))
  (call-in-os-thread
   (λ ()
     (os-async-channel-put os-ach
                           `(Done! ,(libc:sleep seconds)))))
  (sync os-ach))
(demo 5)
```
